ordered_set.py

```diff
@@ -509,3 +507,285 @@ def symmetric_difference_update(self, other: SetLike[T]) -> None:
         self._update_items(
             [item for item in self.items if item not in items_to_remove] + items_to_add
         )
+
+
+class FrozenOrderedSet(AbstractSet[T], Sequence[T]):
+    """
+    A FrozenOrderedSet is a custom AbstractSet that remembers its order, so that
+    every entry has an index that can be looked up.
+
+    Example:
+        >>> FrozenOrderedSet([1, 1, 2, 3, 2])
+        FrozenOrderedSet([1, 2, 3])
+    """
+
+    def __init__(self, iterable: Optional[Iterable[T]] = None):
+        self.items = tuple(OrderedSet(iterable))  # type: Tuple[T, ...]
```

Comment:
This line is key. It allows us to avoid having to recreate the ordered-set-like functionality of `OrderedSet` but saves the values in an immutable tuple so that it's frozen.